server/state/in_memory_state.go

```diff
@@ -0,0 +1,460 @@
+	lock         sync.Mutex
+	sandboxes    map[string]*sandbox.Sandbox
+	containers   oci.ContainerStorer
+	podNameIndex *registrar.Registrar
+	podIDIndex   *truncindex.TruncIndex
+	ctrNameIndex *registrar.Registrar
+	ctrIDIndex   *truncindex.TruncIndex
+}
+
+// NewInMemoryState creates a new, empty server state
+func NewInMemoryState() RuntimeStateStorer {
+	state := new(InMemoryState)
+	state.sandboxes = make(map[string]*sandbox.Sandbox)
+	state.containers = oci.NewMemoryStore()
+	state.podNameIndex = registrar.NewRegistrar()
+	state.podIDIndex = truncindex.NewTruncIndex([]string{})
+	state.ctrNameIndex = registrar.NewRegistrar()
+	state.ctrIDIndex = truncindex.NewTruncIndex([]string{})
+
+	return state
+}
+
+// AddSandbox adds a sandbox and any containers in it to the state
+func (s *InMemoryState) AddSandbox(sandbox *sandbox.Sandbox) (err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if sandbox == nil {
+		return fmt.Errorf("nil passed as sandbox to AddSandbox")
+	}
+
+	if _, exist := s.sandboxes[sandbox.ID()]; exist {
+		return fmt.Errorf("sandbox with ID %v already exists", sandbox.ID())
+	}
+
+	// We shouldn't share ID with any containers
+	// Our pod infra container will share our ID and we don't want it to conflict with anything
+	if ctrCheck := s.containers.Get(sandbox.ID()); ctrCheck != nil {
+		return fmt.Errorf("requested sandbox ID %v conflicts with existing container ID", sandbox.ID())
+	}
+
+	if err := s.addSandboxMappings(sandbox); err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			if err2 := s.deleteSandboxMappings(sandbox); err2 != nil {
+				logrus.Errorf("Error removing mappings for incompletely-added sandbox %s: %v", sandbox.ID(), err)
+			}
+		}
+	}()
+
+	// If there are containers in the sandbox add them to the mapping
+	addedCtrs := make([]*oci.Container, 0, len(sandbox.Containers()))
+	for _, ctr := range sandbox.Containers() {
+		if err := s.addContainerMappings(ctr, true); err != nil {
+			return fmt.Errorf("error adding container %v mappings in sandbox %v", ctr.ID(), sandbox.ID())
+		}
+		addedCtrs = append(addedCtrs, ctr)
+	}
+	defer func() {
+		if err != nil {
+			for _, ctr := range addedCtrs {
+				if err2 := s.deleteContainerMappings(ctr, true); err2 != nil {
+					logrus.Errorf("Error removing container %s mappings: %v", ctr.ID(), err2)
+				}
+			}
+		}
+	}()
+
+	// Add the pod infrastructure container to mappings
+	// TODO: Right now, we don't add it to the all containers listing. We may want to change this.
+	if err := s.addContainerMappings(sandbox.InfraContainer(), false); err != nil {
+		return fmt.Errorf("error adding infrastructure container %v to mappings: %v", sandbox.InfraContainer().ID(), err)
+	}
+
+	return nil
+}
+
+// Add sandbox name, ID to appropriate mappings
+func (s *InMemoryState) addSandboxMappings(sb *sandbox.Sandbox) error {
+	if _, exist := s.sandboxes[sb.ID()]; exist {
+		return fmt.Errorf("sandbox with ID %s already exists in sandboxes map", sb.ID())
+	}
```

Comment:
I removed the check from AddSandbox(), given that addSandboxMappings() could be called elsewhere, so it seems the more logical place to keep the check